XML Controller VID / PID definitions

jdfeng opened the pull request:
Hello!

This is the prelimary work to move the VID / PID controller definitions out of Java and into XML that I mentioned in Issue #265 .   It moves the definitions out of DS4Devices.cs and into app.config so it's easier for someone with a release build to try adding their controller and seeing if it works.

Please let me know if you have any questions / comments / feedback!  Thank you so much for improving DS4Windows!

jdfeng commented on DS4Windows/DS4Library/DS4Devices.cs:
```diff
@@ -1,23 +1,13 @@
 
 namespace DS4Windows
 {
-    public class VidPidInfo
```
VidPidInfo was moved into SupportedControllers.cs

jdfeng commented on DS4Windows/DS4Library/DS4Devices.cs:
```diff
@@ -28,16 +18,18 @@ public class DS4Devices
         private static List<HidDevice> DisabledDevices = new List<HidDevice>();
         private static Stopwatch sw = new Stopwatch();
         public static bool isExclusiveMode = false;
-        internal const int SONY_VID = 0x054C;
```
All of these default configurations have been moved into app.config

jdfeng commented on DS4Windows/DS4Library/DS4Devices.cs:
```diff
@@ -28,16 +18,18 @@ public class DS4Devices
-        };
+        private static readonly VidPidInfo[] knownDevices;
+
+        static DS4Devices() {
```
DS4Devices now loads the default controller definitions from app.config.

jdfeng commented on DS4Windows/DS4Library/SupportedControllers.cs:
```diff
@@ -0,0 +1,57 @@
+using System.Globalization;
+
+namespace DS4Windows {
+    public class SupportedControllers : ConfigurationSection {
```
This is the parser for the app.config controller configuration sections.

jdfeng commented on DS4Windows/DS4Windows.csproj:
```diff
@@ -39,6 +39,7 @@
     <ErrorReport>prompt</ErrorReport>
     <WarningLevel>4</WarningLevel>
     <Prefer32Bit>false</Prefer32Bit>
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
```
Not sure if this is appropriate - it was required for me to compile the code locally, as it contained unsafe blocks.

jdfeng commented on DS4Windows/DS4Windows.csproj:
```diff
@@ -124,6 +125,7 @@
       <HintPath>..\packages\TaskScheduler.2.8.0\lib\net452\Microsoft.Win32.TaskScheduler.dll</HintPath>
     </Reference>
     <Reference Include="System" />
+    <Reference Include="System.configuration" />
```
Needed to parse the app.config file during execution.

jdfeng commented on DS4Windows/HidLibrary/HidDevices.cs:
```diff
@@ -17,59 +17,12 @@ public static bool IsConnected(string devicePath)
 
         public static HidDevice GetDevice(string devicePath)
         {
-            return Enumerate(devicePath).FirstOrDefault();
```
Removed the unused methods.

jdfeng commented on DS4Windows/app.config:
```diff
@@ -1,5 +1,8 @@
 <?xml version="1.0" encoding="utf-8"?>
 <configuration>
+    <configSections>
```
Defines the new parser for the SupportedControllers element

jdfeng commented on DS4Windows/app.config:
```diff
@@ -8,4 +11,19 @@
             <probing privatePath="Lang"/>
         </assemblyBinding>
     </runtime>
+    <SupportedControllers>
```
Includes the configurations for all of the existing parsers, plus the Hori Mini from my other pull request

Ryochan7 commented:
The idea is good and devices are matched as they should be based on my testing. However, enough is changed in the code to affect input latency for unknown reasons. I have not played around with changing the branch to see if anything can be done to get rid of the problem. As it stands, the problem would be enough to keep the pull request from getting committed.

jdfeng commented:
I rebased to include your recent commits, reverted the change that I suspect was causing the latency, and also tweaked my VidPidInfo changes to reduce unnecessary recalc.  Please let me know if this fixes the input latency you saw!  And if there's a way for me to test the input latency on my change locally, please let me know too - I'm happy to do it before I make a future pull request, just don't know how :)

Reakleases commented:
bump

Ryochan7 commented:
I deleted the old copy and pulled the latest version of the branch. I also performed a rebase to include some newer commits that are not incorporated in this version. Unfortunately, there was a noticeable different in input delay during play testing with the changes. It also seems to affect the numbers displayed in the Controller Readings tab and overall UI performance; the input delay number is more skewed towards and sometimes past 5 ms and switching tabs and opening the About window takes some extra time to perform.

I had tried to make some minor edits before but I was not successful in affecting the performance. Maybe changing around HidDevice.cs was a contributing factor. The idea behind this pull request is good but the .NET runtime does not seem to like it. At this point, it looks like this change will not get merged.

jdfeng commented:
Hmm, interesting.  I pulled your latest changes, rebased,  set breakpoints in all of the changes I made in HidDevices.cs (which I assume is the file you meant, as my PR doesn't change any files in HidDevice.cs ),  DS4Devices.cs, and VidPidInfo in SupportedControllers.cs (HidDevices.EnumerateDS4(), HidDevices.GetDevice(), DS4Devices' static constructor, and every method and property in VidPidInfo.) All of the code I changed is only executed during application initialization; none of the breakpoints were hit during Controller Readings.  The XML-based list is only loaded into memory once, during static initializaton of DS4Devices.cs (same as the original String-based implementation).

I did observe an initial slowness in the Controller Reading behaviour, but I also saw the same slowness with a clean branch (without my changes).  Input delay is ~ 4ms (occasionally fluctuating from 3.95-4.05ms), but both the delay and the fluctuation occur with and without my changes.  (And none of the breakpoints for my changes are hit).

Ryochan7 commented:
There are some weird inconsistencies with .NET. I have run across several occasions when odd code changes have undesired consequences. Something as simple as flipping the clauses of an if statement, which would result in more work for the .NET Runtime, can affect performance drastically in given situations. The latest blunder that I had encountered was when .NET decided that it needed a function call to be included that was actually redundant and would result in no real work being done. There was a good performance dip by just removing the one call so I had to add it back.

https://github.com/Ryochan7/DS4Windows/commit/a6b9fb0f4e4f07fb8887335e98797f89cd91a69a

There are a couple of classes still included in the project despite not being used anywhere in the project due to .NET making DS4Windows run worse without them; DS4Keyboard and AlreadyRunning are the examples that I am thinking of.

jdfeng commented:
Thank you!   

I've pared down the number of properties stored in VidPidInfo, and also reached out to a friend who's an expert on C# .NET (as in helped to write the framework).  My friend mentioned the following:

1. My code was adding a static constructor (instead of a field initiliazation) in DS4Devices.cs for knownDevices, and that could impact performance.  I've refactored that in 3444db9 
2. Deleting unused methods and classes should have no impact on performance: 
> Unused code in classes that are used gets loaded, but unused code in unused classes doesn't even get looked at.  Unused code isn't going to change stack alignment

I wish I had a better way to benchmark perf issues; unfortunately I can't measure any distinct differences between the baseline and my changes.  My friend also mentioned that there are differences between the way .NET is handled on different processors (ex: Intel vs AMD), which may be why we see different results.  I hope this set of changes is a bit more performant though.

If it helps:

I tried locally deleting DS4Keyboard and AlreadyRunning, and didn't notice any perf differences (still observed slowness in the Controller Reading tab for the first 50s, but switching tabs and accessing 'About' still were responsive, and input delay was consistent at 4ms.)

The one-call removal in a6b9fb0 looks like it may be needed because that StopOutputUpdate(); doesn't get run by default for all calls to sendOutputReport(true, true); sendOutputReport(true, true) only calls StopOutputUpdate() when writeOutput() fails with an error).  Although removing that still shouldn't have affected runtime performance, just controller disconnect / reconnect...  (Although perhaps with the recent rise of disconnect / reconnect issues  reported, that code may be called more often?  And because removing it changes the stop basically to a weird timeout state, it's perceived as a perf degradation?)

Ryochan7 commented:
I performed a fresh checkout of this branch and I resolved the current conflicts to try it out against the current code base. Unfortunately, it is still a no go. GUI and output mapping performance are definitely worse with the changes enabled. At this point, I am not going to take time to try to edit the changes again to attempt to improve application performance. I will consider this a lost cause and continue using the current method of defining supported controllers. I am sorry for not incorporating these changes.

On a related note, I had a major problem happen recently while attempting to edit auto profile functionality. Changing just a few lines of code ended up taking several extra hours just to get .NET to cooperate.

https://github.com/Ryochan7/DS4Windows/commit/284011c913f61b95bcfdf822cc86b1dc851db4d0
